### PR TITLE
feat: create erc20 contract on denom metadata registration

### PR DESCRIPTION
**Pull request by zale144**
## Description

----

Closes #dymension-rdk/311

### CODE REVIEW

File: x/erc20/keeper/denom_metadata_hooks.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+func (e ERC20BankContractRegistrationHook) AfterDenomMetadataCreation(ctx sdk.Context, newDenomMetadata banktypes.Metadata) error {
+	// only deploy for IBC denom.
+	if !strings.HasPrefix(strings.ToLower(newDenomMetadata.Base), "ibc/") {
+		return nil
+	}
+
+	if _, err := e.erc20Keeper.RegisterCoin(ctx, newDenomMetadata); err != nil {
+		return fmt.Errorf("failed to deploy the erc20 contract for the IBC coin: %s; error: %w", newDenomMetadata.Base, err)
```
Comment (danwt) [inline x/erc20/keeper/denom_metadata_hooks.go]:
```suggestion
		return fmt.Errorf("deploy the erc20 contract for the ibc coin: %s; error: %w", newDenomMetadata.Base, err)
```

Comment (danwt) [reply x/erc20/keeper/denom_metadata_hooks.go]: avoiding caps

Comment (danwt) [reply x/erc20/keeper/denom_metadata_hooks.go]: don't need 'failed to..'

File: x/erc20/keeper/ibc_callbacks.go [hunk 2]
```diff
@@ -101,15 +100,25 @@ func (k Keeper) OnRecvPacket(
...
-	if !pair.Enabled {
+	pair, found := k.GetTokenPair(ctx, pairID)
+	if !(found && pair.Enabled) {
 		// no-op: continue with the rest of the stack without conversion
 		return ack
 	}
 
+	address := common.HexToAddress(pair.Erc20Address)
+	if !k.IsERC20Registered(ctx, address) {
+		err := errorsmod.Wrapf(types.ErrERC20TokenPairNotRegistered, "ERC20 token pair %s does not exist", coin.Denom)
```
Comment (danwt) [inline x/erc20/keeper/ibc_callbacks.go]:
```suggestion
		err := errorsmod.Wrapf(types.ErrERC20TokenPairNotRegistered, "erc20 token pair %s does not exist", coin.Denom)
```

File: x/erc20/keeper/proposals.go [hunk 3]
```diff
@@ -32,20 +30,19 @@ func (k Keeper) RegisterCoin(
...
 			types.ErrTokenPairAlreadyExists, "coin denomination already registered: %s", coinMetadata.Name,
 		)
 	}
 
-	// Check if the coin exists by ensuring the supply is set
-	if !k.bankKeeper.HasSupply(ctx, coinMetadata.Base) {
-		return nil, errorsmod.Wrapf(
-			errortypes.ErrInvalidCoins, "base denomination '%s' cannot have a supply of 0", coinMetadata.Base,
-		)
-	}
```
Comment (danwt) [inline x/erc20/keeper/proposals.go:47]: I think we should have a comment explaining why this is OK to delete

File: x/erc20/keeper/proposals.go [hunk 4]
```diff
@@ -32,20 +30,19 @@ func (k Keeper) RegisterCoin(
 	ctx sdk.Context,
 	coinMetadata banktypes.Metadata,
 ) (*types.TokenPair, error) {
+	// Check if ERC20 is enabled
+	if !k.IsERC20Enabled(ctx) {
+		return nil, errorsmod.Wrap(
+			types.ErrERC20Disabled, "ERC20 is disabled",
```
Comment (danwt) [inline x/erc20/keeper/proposals.go]:
```suggestion
			types.ErrERC20Disabled, "erc20 is disabled",
```

File: x/erc20/keeper/ibc_callbacks_test.go [hunk 5]
```diff
@@ -372,6 +389,10 @@ func (suite *KeeperTestSuite) TestOnRecvPacket() {
 				suite.Require().NoError(err)
 			}
 
+			if tc.deleteERC20Token != "" {
+				suite.app.Erc20Keeper.DeleteERC20Map(suite.ctx, common.HexToAddress(tc.deleteERC20Token))
+			}
```
Comment (danwt) [inline x/erc20/keeper/ibc_callbacks_test.go]: why is this necessary?

File: x/erc20/keeper/ibc_callbacks_test.go [hunk 6]
```diff
@@ -246,6 +247,22 @@ func (suite *KeeperTestSuite) TestOnRecvPacket() {
 			expErc20s:     big.NewInt(0),
 			expCoins:      coins,
 		},
+		{
+			name: "error - erc20 token not registered",
```
Comment (danwt) [inline x/erc20/keeper/ibc_callbacks_test.go]: in the name can you put why it won't succeed

File: x/erc20/keeper/denom_metadata_hooks.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	}
+
+	return nil
+}
+
+func (e ERC20BankContractRegistrationHook) AfterDenomMetadataUpdate(sdk.Context, banktypes.Metadata) error {
+	// do nothing
```
Comment (danwt) [inline x/erc20/keeper/denom_metadata_hooks.go]: is it really fine to do nothing? on the face of it, we'd need to update the contract or something

File: x/erc20/keeper/ibc_callbacks.go [hunk 8]
```diff
@@ -101,15 +100,25 @@ func (k Keeper) OnRecvPacket(
...
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
 	// Instead of converting just the received coins, convert the whole user balance
 	// which includes the received coins.
 	balance := k.bankKeeper.GetBalance(ctx, recipient, coin.Denom)
+	if !balance.IsPositive() {
+		// no-op: continue with the rest of the stack without conversion
+		return ack
+	}
```
Comment (danwt) [inline x/erc20/keeper/ibc_callbacks.go]: can the comment say why, not what

File: x/erc20/keeper/denom_metadata_hooks.go [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+// NewERC20ContractRegistrationHook returns the DenomMetadataHooks for ERC20 registration
+func NewERC20ContractRegistrationHook(erc20Keeper Keeper) ERC20BankContractRegistrationHook {
+	return ERC20BankContractRegistrationHook{
+		erc20Keeper: erc20Keeper,
+	}
+}
+
+func (e ERC20BankContractRegistrationHook) AfterDenomMetadataCreation(ctx sdk.Context, newDenomMetadata banktypes.Metadata) error {
```
Comment (danwt) [inline x/erc20/keeper/denom_metadata_hooks.go:32]: I think it would be good to explain the whole feature in a docstring on this function

To make sure I understand, is the following text correct?

```
We assume that whoever is sending IBC packets with a token denom X, already did a CreateDenomMetadata for that token. At that time, the hook will be fired and the contract will have been created.

Therefore, when the packet arrives, the contract is already created, and we can convert to erc20 freely
```

I still don't understand why we have the positive balance check.

Also, is something needed for the reverse direction, erc20->ibc?

**Review by danwt — CHANGES_REQUESTED** [review]
Hey Alex

Looks like nice work

I'm think I'm being dumb as I'm super tired atm, but I'm having trouble fully understanding

Most of my comments are asking for more comments and docstrings

File: x/erc20/keeper/proposals.go (hunk 3 above)
Comment (zale144) [reply x/erc20/keeper/proposals.go:47]: leave a comment in place of the code removed? I guess it can say something like "No need to check if the coin exists here"

File: x/erc20/keeper/ibc_callbacks_test.go (hunk 5 above)
Comment (zale144) [reply x/erc20/keeper/ibc_callbacks_test.go]: because we should test what happens when there is no ERC20 contract for the denom.

I agree it could have been a bit more elegant...

File: x/erc20/keeper/ibc_callbacks_test.go (hunk 6 above)
Comment (zale144) [reply x/erc20/keeper/ibc_callbacks_test.go]: But it says it already - It will error because the erc20 token was not registered prior to the IBC transfer

File: x/erc20/keeper/denom_metadata_hooks.go (hunk 7 above)
Comment (zale144) [reply x/erc20/keeper/denom_metadata_hooks.go]: Can we do that? Smart contracts are immutable by design

File: x/erc20/keeper/ibc_callbacks.go (hunk 8 above)
Comment (zale144) [reply x/erc20/keeper/ibc_callbacks.go]: it's just my assumption here, if the user has no coins in the account, they can't do a transfer

File: x/erc20/keeper/denom_metadata_hooks.go (hunk 9 above)
Comment (zale144) [reply x/erc20/keeper/denom_metadata_hooks.go:32]: > We assume that whoever is sending IBC packets with a token denom X, already did a CreateDenomMetadata for that token.

I don't think it has to be the same user. In order to create the denom, you need to have a special permission. To do an IBC transfer, you don't need it.

But the idea is more less that, first register the denom in the bank, and in the erc20 module (with the contract), and only after that you can successfully make IBC transfers for that denom. If the denom (metadata) is missing in the bank, or the contract is not created for whatever reason, the IBC transfer should fail.

The positive balance check is just a courtesy check I thought might be nice, so we don't go any further when there's no balance on the user's account. I'm not 100% sure we need it though

Comment (zale144) [reply x/erc20/keeper/denom_metadata_hooks.go:32]: I do agree it should be documented. 

As for the erc20-> ibc, that's a good question. For @omritoptix. Because I have no idea

File: x/erc20/keeper/denom_metadata_hooks.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+		erc20Keeper: erc20Keeper,
+	}
+}
+
+// AfterDenomMetadataCreation deploys the ERC20 contract for the IBC denom. It is called after the
+// bank module creates the denom metadata. Without the ERC20 contract, the IBC denom cannot be
+// converted to the ERC20 token, and the IBC transfer will fail.
+func (e ERC20BankContractRegistrationHook) AfterDenomMetadataCreation(ctx sdk.Context, newDenomMetadata banktypes.Metadata) error {
+	// only deploy for IBC denom.
+	if !strings.HasPrefix(strings.ToLower(newDenomMetadata.Base), "ibc/") {
```
Comment (mtsitrin) [inline x/erc20/keeper/denom_metadata_hooks.go]: pls keep the `ibc/` string in a variable

File: x/erc20/keeper/denom_metadata_hooks.go (hunk 7 above)
Comment (mtsitrin) [reply x/erc20/keeper/denom_metadata_hooks.go]: I suggest returning an error here if trying to update a metadata with an already existing contract.
It should fail the update tx as well.
I think it's better than the alternative, where we have difference between the bank module and the erc20 module

File: x/erc20/keeper/denom_metadata_hooks.go (hunk 9 above)
Comment (mtsitrin) [reply x/erc20/keeper/denom_metadata_hooks.go:32]: once erc20 contract exists, the conversion is automatic.
to create ERC20 contract for IBC tokens, this PR handles it
to turn a new ERC20 to IBC, a `RegisterERC20` should be called, which is currently triggered by gov only.

File: x/erc20/keeper/ibc_callbacks.go [hunk 11]
```diff
@@ -101,15 +100,25 @@ func (k Keeper) OnRecvPacket(
 		return ack
 	}
 
-	pair, _ := k.GetTokenPair(ctx, pairID)
-	if !pair.Enabled {
+	pair, found := k.GetTokenPair(ctx, pairID)
```
Comment (mtsitrin) [inline x/erc20/keeper/ibc_callbacks.go]: @zale144 won't it fail on `GetTokenPairID` before if not registered?

File: x/erc20/keeper/ibc_callbacks.go [hunk 12]
```diff
@@ -101,15 +100,25 @@ func (k Keeper) OnRecvPacket(
...
+	if !(found && pair.Enabled) {
 		// no-op: continue with the rest of the stack without conversion
 		return ack
 	}
 
+	address := common.HexToAddress(pair.Erc20Address)
```
Comment (mtsitrin) [inline x/erc20/keeper/ibc_callbacks.go]: not sure why u need this code
already called in `GetTokenPairID`

File: x/erc20/keeper/ibc_callbacks.go [hunk 13]
```diff
@@ -101,15 +100,25 @@ func (k Keeper) OnRecvPacket(
...
+		err := errorsmod.Wrapf(types.ErrERC20TokenPairNotRegistered, "erc20 token pair %s does not exist", coin.Denom)
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
 	// Instead of converting just the received coins, convert the whole user balance
 	// which includes the received coins.
 	balance := k.bankKeeper.GetBalance(ctx, recipient, coin.Denom)
+	if !balance.IsPositive() {
+		// if the balance is zero, no need to convert
+		return ack
```
Comment (mtsitrin) [inline x/erc20/keeper/ibc_callbacks.go]: why not letting the convert flow run and fail? why explicitly check for balance here

File: x/erc20/keeper/ibc_callbacks.go (hunk 12 above)
Comment (zale144) [reply x/erc20/keeper/ibc_callbacks.go]: It's called above by passing the `coin.Denom` which is not the token address, like here I think. Here we explicitly check if the token is stored by the address

File: x/erc20/keeper/ibc_callbacks.go (hunk 11 above)
Comment (zale144) [reply x/erc20/keeper/ibc_callbacks.go]: not necessarily fail, more like return early. I guess it would, even though those are separate stores. the extra `found` check is probably redundant.

File: x/erc20/keeper/denom_metadata_hooks.go [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+	if _, err := e.erc20Keeper.RegisterCoin(ctx, newDenomMetadata); err != nil {
+		return fmt.Errorf("deploy the erc20 contract for the ibc coin: %s; error: %w", newDenomMetadata.Base, err)
+	}
+
+	return nil
+}
+
+func (e ERC20BankContractRegistrationHook) AfterDenomMetadataUpdate(sdk.Context, banktypes.Metadata) error {
+	return fmt.Errorf("update the denom metadata while having an already existing contract")
```
Comment (hoangdv2429) [inline x/erc20/keeper/denom_metadata_hooks.go:46]: I'm not really sure why we have the hook this way. Does that mean that we can't upgrade denommetadata of an existing erc20 token ?

Comment (zale144) [reply x/erc20/keeper/denom_metadata_hooks.go:46]: if it's an ERC20 token, we need to implement the hook in the `x/erc20` module, which would not support the metadata update. But we still need to implement the interface. if it was handled by a different module, we still need to leave the possibility of handling the update flow hook

**Review by mtsitrin — APPROVED** [review]

**Review by omritoptix — APPROVED** [review]

File: x/erc20/keeper/denom_metadata_hooks_test.go [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			e := keeper.NewERC20ContractRegistrationHook(tt.fields.erc20Keeper)
+			err := e.AfterDenomMetadataCreation(sdk.Context{}, tt.args.newDenomMetadata)
+			if tt.expectErr != "" {
+				require.EqualError(t, err, tt.expectErr)
+			} else {
+				require.NoError(t, err)
+			}
```
Comment (danwt) [inline x/erc20/keeper/denom_metadata_hooks_test.go:77]: can we replace the errors with Err* types and use require.True(t, errors.Is(err, ErrFoo))

Comment (danwt) [reply x/erc20/keeper/denom_metadata_hooks_test.go:77]: aka no string matching plz 🙏

File: x/erc20/keeper/ibc_callbacks.go [hunk 16]
```diff
@@ -96,9 +99,8 @@ func (k Keeper) OnRecvPacket(
 
 	pairID := k.GetTokenPairID(ctx, coin.Denom)
 	if len(pairID) == 0 {
-		// short-circuit: if the denom is not registered, conversion will fail
-		// so we can continue with the rest of the stack
-		return ack
+		err = errorsmod.Wrapf(types.ErrTokenPairNotFound, "token pair not found for %s", coin.Denom)
```
Comment (danwt) [inline x/erc20/keeper/ibc_callbacks.go:102]:
```suggestion
		err = errorsmod.Wrapf(types.ErrTokenPairNotFound, "token pair not found: denom: %s", coin.Denom)
```

File: x/erc20/keeper/proposals.go (hunk 3 above)
Comment (danwt) [reply x/erc20/keeper/proposals.go:47]: I mean, why did we need to check before, but not now?
Like, why did the check exist before?

**Review by danwt — APPROVED** [review]

**Comment by danwt** [comment]
Approved but I left an improvement for the test

File: x/erc20/keeper/denom_metadata_hooks_test.go (hunk 15 above)
Comment (zale144) [reply x/erc20/keeper/denom_metadata_hooks_test.go:77]: addressed here: https://github.com/dymensionxyz/evmos/pull/23

File: x/erc20/keeper/ibc_callbacks.go (hunk 16 above)
Comment (zale144) [reply x/erc20/keeper/ibc_callbacks.go:102]: addressed here: https://github.com/dymensionxyz/evmos/pull/23
